File: web/src/api/quota-settings.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { z } from "zod";
import { apiClient } from "@/api/client";

// 백엔드 quota_settings 한 행 - GET/PATCH 모두 이 배열을 그대로 반환한다.
export const QuotaSettingSchema = z.object({
  key: z.string(),
  value: z.string(),
  updated_at: z.string(),
  updated_by: z.string().nullable(),
});
export type QuotaSetting = z.infer<typeof QuotaSettingSchema>;

// 화이트리스트 key → 라벨·검증 범위 (백엔드 core.quota_settings.QUOTA_SETTING_RULES와 일치).
// 조직 상한은 1~1,000,000, 역할 기본은 1~100,000.
export const QUOTA_META = [
  {
    key: "ORG_MONTHLY_COST_LIMIT_USD",
    label: "조직 전체 월 한도",
    hint: "회사 전체 월 비용 천장 - 초과 시 전원 차단",
    min: 1,
    max: 1_000_000,
  },
  {
    key: "DEFAULT_LIMIT_SUPER_ADMIN_USD",
    label: "super_admin 기본 한도",
    hint: "개인 지정이 없는 super_admin에게 적용",
    min: 1,
    max: 100_000,
  },
  {
    key: "DEFAULT_LIMIT_ADMIN_USD",
    label: "admin 기본 한도",
    hint: "개인 지정이 없는 admin에게 적용",
    min: 1,
    max: 100_000,
  },
  {
    key: "DEFAULT_LIMIT_WORKER_USD",
    label: "worker 기본 한도",
    hint: "개인 지정이 없는 worker(신규 포함)에게 적용",
    min: 1,
    max: 100_000,
  },
  // viewer 한도 설정란은 없음 - 뷰어는 읽기 전용이라 기본 $0 고정(2026-08-15 결정).
  // LLM 호출 경로 자체가 없어 조정할 이유가 없고, 조정이 필요해지면 그때 역할 설계부터
  // 다시 본다(백엔드 core.limit DEFAULT_ROLE_LIMIT_USD.viewer = 0).
] as const;

export const quotaSettingsKeys = {
  all: ["quota-settings"] as const,
};

export async function getQuotaSettings(): Promise<QuotaSetting[]> {
  const data = await apiClient.get<unknown>("admin/quota-settings");
  return z.array(QuotaSettingSchema).parse(data);
}

export function useQuotaSettings() {
  return useQuery({ queryKey: quotaSettingsKeys.all, queryFn: getQuotaSettings });
}

// 배치 수정 - {key: value} 딕셔너리. 백엔드가 검증·감사 후 전체 최신본을 돌려준다.
export async function updateQuotaSettings(patch: Record<string, string>): Promise<QuotaSetting[]> {
  const data = await apiClient.patch<unknown>("admin/quota-settings", { json: patch });
  return z.array(QuotaSettingSchema).parse(data);
}

export function useUpdateQuotaSettings() {
  const qc = useQueryClient();
  return useMutation({
    mutationKey: [...quotaSettingsKeys.all, "update"],
    mutationFn: updateQuotaSettings,
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: quotaSettingsKeys.all });
    },
  });
}

File: src/core/quota_settings.py
from enum import StrEnum
from typing import NamedTuple


class QuotaSettingKey(StrEnum):
    """
    quota_settings.key 화이트리스트

    기존 config 상수(Settings.org_monthly_cost_limit_usd,
    core.limit.DEFAULT_ROLE_LIMIT_USD)를 대체한다. 여기 없는 key는
    조회/수정 모두 거부된다.
    """

    ORG_MONTHLY_COST_LIMIT_USD = "ORG_MONTHLY_COST_LIMIT_USD"
    DEFAULT_LIMIT_SUPER_ADMIN_USD = "DEFAULT_LIMIT_SUPER_ADMIN_USD"
    DEFAULT_LIMIT_ADMIN_USD = "DEFAULT_LIMIT_ADMIN_USD"
    DEFAULT_LIMIT_WORKER_USD = "DEFAULT_LIMIT_WORKER_USD"
    # viewer 한도는 설정 대상이 아니다 — 읽기 전용 역할이라 LLM 비용 경로가 없어
    # $0 고정(core.limit). 여기서 빠졌으므로 조회 캐시·수정 모두 거부된다(0040에서
    # 기존 시딩 행 삭제). 조정이 필요해지면 역할 설계부터 다시 본다(2026-08-15 결정).


class QuotaSettingRule(NamedTuple):
    min_value: int
    max_value: int
    description: str


# key별 값 검증 규칙 — 전부 "양의 정수(USD)" 도메인이므로 정수 파싱 + 범위 검증만 수행한다.
QUOTA_SETTING_RULES: dict[QuotaSettingKey, QuotaSettingRule] = {
    QuotaSettingKey.ORG_MONTHLY_COST_LIMIT_USD: QuotaSettingRule(
        1, 1_000_000, "조직 전체 월 비용 한도(USD)"
    ),
    QuotaSettingKey.DEFAULT_LIMIT_SUPER_ADMIN_USD: QuotaSettingRule(
        1, 100_000, "super_admin 역할 기본 월 한도(USD)"
    ),
    QuotaSettingKey.DEFAULT_LIMIT_ADMIN_USD: QuotaSettingRule(
        1, 100_000, "admin 역할 기본 월 한도(USD)"
    ),
    QuotaSettingKey.DEFAULT_LIMIT_WORKER_USD: QuotaSettingRule(
        1, 100_000, "worker 역할 기본 월 한도(USD)"
    ),
}

# 마이그레이션 초기 시딩값 — 이관 전 config 상수의 실제 기본값을 그대로 옮긴다.
# (Settings.org_monthly_cost_limit_usd=3000, core.limit.DEFAULT_ROLE_LIMIT_USD)
QUOTA_SETTING_SEED_VALUES: dict[str, str] = {
    QuotaSettingKey.ORG_MONTHLY_COST_LIMIT_USD.value: "3000",
    QuotaSettingKey.DEFAULT_LIMIT_SUPER_ADMIN_USD.value: "500",
    QuotaSettingKey.DEFAULT_LIMIT_ADMIN_USD.value: "300",
    QuotaSettingKey.DEFAULT_LIMIT_WORKER_USD.value: "200",
}


# 역할 → 기본 한도 quota_settings key 매핑. core.limit.DEFAULT_ROLE_LIMIT_USD와 동일한
# 역할 집합을 다룬다. 알 수 없는 역할은 매핑에 없음(quota_key_for_role은 None 반환).
QUOTA_KEY_FOR_ROLE: dict[str, QuotaSettingKey] = {
    "super_admin": QuotaSettingKey.DEFAULT_LIMIT_SUPER_ADMIN_USD,
    "admin": QuotaSettingKey.DEFAULT_LIMIT_ADMIN_USD,
    "worker": QuotaSettingKey.DEFAULT_LIMIT_WORKER_USD,
    # viewer는 매핑 없음 → quota_key_for_role이 None → core.limit 상수($0) 폴백.
}


def quota_key_for_role(role: str) -> QuotaSettingKey | None:
    """역할에 대응하는 quota_settings key. 알 수 없는 역할이면 None."""
    return QUOTA_KEY_FOR_ROLE.get(role)


def parse_quota_setting_key(key: str) -> QuotaSettingKey:
    """key가 화이트리스트에 있는지 확인한다. 없으면 ValueError."""
    try:
        return QuotaSettingKey(key)
    except ValueError:
        raise ValueError(f"화이트리스트에 없는 key입니다: {key!r}") from None


def validate_quota_setting_value(key: QuotaSettingKey, value: str) -> int:
    """value가 key의 검증 규칙(정수·범위)을 만족하는지 확인하고 파싱된 정수를 반환한다."""
    try:
        parsed = int(value)
    except (TypeError, ValueError):
        raise ValueError(f"'{key.value}' 값은 정수여야 합니다: {value!r}") from None

    rule = QUOTA_SETTING_RULES[key]
    if not (rule.min_value <= parsed <= rule.max_value):
        raise ValueError(
            f"'{key.value}' 값은 {rule.min_value}~{rule.max_value} 범위여야 합니다: {parsed}"
        )
    return parsed
